processor/intervalprocessor/processor.go

```diff
@@ -52,5 +79,185 @@ func (p *Processor) Capabilities() consumer.Capabilities {
 }
 
 func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
-	return p.nextConsumer.ConsumeMetrics(ctx, md)
+	var errs error
+
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	md.ResourceMetrics().RemoveIf(func(rm pmetric.ResourceMetrics) bool {
+		rm.ScopeMetrics().RemoveIf(func(sm pmetric.ScopeMetrics) bool {
+			sm.Metrics().RemoveIf(func(m pmetric.Metric) bool {
+				switch m.Type() {
+				case pmetric.MetricTypeGauge, pmetric.MetricTypeSummary:
+					return false
+				case pmetric.MetricTypeSum:
+					sum := m.Sum()
+
+					if !sum.IsMonotonic() {
+						return false
+					}
+
+					if sum.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					aggregateDataPoints(sum.DataPoints(), p.numbers, rm.Resource(), rm.SchemaUrl(), sm.Scope(), sm.SchemaUrl(), m)
+					return true
+				case pmetric.MetricTypeHistogram:
+					histogram := m.Histogram()
+
+					if histogram.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					aggregateDataPoints(histogram.DataPoints(), p.histograms, rm.Resource(), rm.SchemaUrl(), sm.Scope(), sm.SchemaUrl(), m)
+					return true
+				case pmetric.MetricTypeExponentialHistogram:
+					expHistogram := m.ExponentialHistogram()
+
+					if expHistogram.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					aggregateDataPoints(expHistogram.DataPoints(), p.expHistograms, rm.Resource(), rm.SchemaUrl(), sm.Scope(), sm.SchemaUrl(), m)
+					return true
+				default:
+					errs = errors.Join(fmt.Errorf("invalid MetricType %d", m.Type()))
+					return false
+				}
+			})
+			return sm.Metrics().Len() == 0
+		})
+		return rm.ScopeMetrics().Len() == 0
+	})
+
+	if err := p.nextConsumer.ConsumeMetrics(ctx, md); err != nil {
+		errs = errors.Join(errs, err)
+	}
+
+	return errs
+}
+
+func aggregateDataPoints[DPS metrics.DataPointSlice[DP], DP metrics.DataPoint[DP]](dataPoints DPS, state map[identity.Stream]metrics.StreamDataPoint[DP], res pcommon.Resource, resSchemaURL string, scope pcommon.InstrumentationScope, scopeSchemaURL string, m pmetric.Metric) {
+	metric := metrics.From(res, resSchemaURL, scope, scopeSchemaURL, m)
+	metricID := metric.Identity()
+
+	now := time.Now()
+
+	for i := 0; i < dataPoints.Len(); i++ {
+		dp := dataPoints.At(i)
+
+		streamDataPointID := metrics.StreamDataPointIdentity(metricID, dp)
+
+		existing, ok := state[streamDataPointID]
+		if !ok {
+			state[streamDataPointID] = metrics.StreamDataPoint[DP]{
+				Metric:      metric,
+				DataPoint:   dp,
+				LastUpdated: now,
+			}
+			continue
+		}
+
+		// Check if the datapoint is newer
+		if dp.Timestamp().AsTime().After(existing.DataPoint.Timestamp().AsTime()) {
```

Comment:
`pcommon.Timestamp` is a uint64, plain comparison would be more efficient here as it does not allocate

Reply:
Good point